components/core/src/tls/ctl_gateway.rs

```diff
@@ -0,0 +1,95 @@
+//! Utilities for generating and reading the self-signed certificate for use with the control
+//! gateway.
+
+use crate::{crypto::keys::NamedRevision,
+            tls::rustls_wrapper::{self,
+                                  Error as RustlsReadersError}};
+use rcgen::{Certificate as RcgenCertificate,
+            CertificateParams,
+            DistinguishedName,
+            DnType,
+            RcgenError,
+            PKCS_ECDSA_P256_SHA256};
+use rustls::{Certificate,
+             PrivateKey,
+             RootCertStore};
+use std::{fs::{self,
+               File},
+          io::{Error as IoError,
+               Write},
+          path::{Path,
+                 PathBuf}};
+use thiserror::Error;
+
+const NAME_PREFIX: &str = "ctl-gateway";
+const CRT_EXTENSION: &str = "crt.pem";
+const KEY_EXTENSION: &str = "key.pem";
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("ctl gateway certificate lookup failed when trying to match files {0}")]
+    FailedToMatchPattern(String),
+    #[error("ctl gateway certificate lookup failed, err: {0}")]
+    RustlsReaders(#[from] RustlsReadersError),
+    #[error("ctl gateway certificate generation failed, err: {0}")]
+    CertificateGeneration(#[from] RcgenError),
+    #[error("writing the ctl gateway certificate failed, err: {0}")]
+    CertificateWrite(#[from] IoError),
+}
+
+pub fn generate_self_signed_certificate_and_key(subject_alternate_name: &str,
+                                                path: impl AsRef<Path>)
+                                                -> Result<(), Error> {
+    let mut params = CertificateParams::new(vec![subject_alternate_name.to_string(),
+                                                 "localhost".to_string(),
+                                                 "127.0.0.1".to_string(),
```

Comment:
I'm not sure it will matter for our use here, but one strange thing about this library is that it assumes all subject alt names are dnsNames:

```
          X509v3 Subject Alternative Name: 
                DNS:foo, DNS:localhost, DNS:127.0.0.1, DNS:0.0.0.0
```

Ideally, IP addresses would get ip address entries there.

Reply:
Thanks, nice catch! I should remove the IP addresses. The `tokio_rustls` library does not allow using an IP address when [specifying the `domain`](https://github.com/habitat-sh/habitat/blob/28bd76a886c6d0f69fd438f6f6754e889ff7d695/components/core/src/tls/rustls_wrapper/tcp_or_tls_stream.rs#L73). As a result, adding those IP addresses do not buy us anything.

Reply:
I imagine the implication then is also that users who want to use this feature for remote supervisors will have to use a hostname rather than an IP to connect.

Reply:
Agreed, that is a nice feature. The field `ctl_server_name_indication` in `cli.toml` allows setting this.